sdk/core/Azure.Core/src/AzureKeyCredential.cs

```diff
@@ -25,6 +26,12 @@ public string Key
             private set => Volatile.Write(ref _key, value);
         }
 
+        /// <summary>
+        /// The event is raised after the AzureKeyCredential.Update method updates the key.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public event EventHandler<string> KeyChanged;
```

Comment:
Would the usage cause a memory leak? Clients are not disposable so would never unsubscribe.

Reply:
Even if they were disposable, it would not solve the problem. Users would not dispose, and we indeed would have a leak. Hmmm, I am not sure what to do about it. The initial idea for this feature was to expose KeyBytes property. But, it also has significant issues:

1. The property (KeyBytes) would need to be synchronized with Key
2. Clients could not still cache things based on the Key unless they always compared the key.
3. Services use different strategies for converting from string key to byte[], e.g. Base64 or Ascii.

Reply:
Consider a `Version` property that would increment every time the key is changes and can be used to cheaply check for validity of the cache.

Reply:
Such version property would need to be synchronized with the key and the bytes. This means a lock around get_Key, get_Bytes, get_Version. Otherwise calling clients could "miss" a change and never recover.

Reply:
No, we can add an atomic deconstruct like we did for AzureNamedKeyCredential policy.

Reply:
If what you are saying is that we don't have KeyBytes and we just return a tuple of Key and Version, then we don't need this API. Clients could as well cache the reference to the Key and compare the reference to decide if they need to regenerate the bytes. But then, the whole race issue is just kicked down the road to user code.

Reply:
>Clients could as well cache the reference to the Key and compare the reference to decide if they need to regenerate the bytes. 

True.

>But then, the whole race issue is just kicked down the road to user code

What is racing with what in that case?

Reply:
I was thinking that concurrent updates to the key value and key version can cause updates to the cached version and the cached bytes to get out of sync. But maybe if I eliminate the version and and just rely on Key reference to compare "versions", it will all be sound. Let me try it.

Reply:
So, I tried the following in the policy:
```csharp
            var key = _credential.Key;
            if (ReferenceEquals(key, _key))
            {
                _keyBytes = Encoding.UTF8.GetBytes(key);
                _key = key;
            }
```
I think the problem is still there. The most severe problem is that the policy end up with a new _key and old _keyBytes. In such case the bytes might never be updated and when the key expires, the policy would stop working. I think this can happen if the CPU reorders the stores to _keyBytes and _key. i.e. one thread could win setting the key and one could win setting the bytes.

Reply:
Just to confirm. This has a problem as well. I will close this draft PR and implement the cache in the policy. The cache will rely on not having any shared state that cannot be updated atomically (i.e. a pointer).